### PR TITLE
[0D-Tensor] Add ExpandZeroDimPass, support 0D input and output

### PR DESCRIPTION
**Pull request by jiahy0825**
[0D-Tensor] Add ExpandZeroDimPass, support 0D input and output

**Note: What does ExpandZeroDimPass do?**

Expand all 0D-Tensor (both inputs and outputs) to 1D-Tensor. To change the 0D-Tensor back when we fetch outputs, this PR records the 0D outputs Tensor's name before `ExpandZeroDimPass`.

**Note: Why register program pass instead of graph pass?**

In the beginning, I try to register ExpandZeroDimPass as the first graph pass to handle 0D-Tensor, however, before the graph passes, the program passes such as `Decomposer` will change instructions with 0D-Tensor input. 

Such change is not what we want, and adapting the program pass with `ExpandZeroDimPass` will introduce a lot of redundant codes which couple program pass and graph pass. So I register `ExpandZeroDimPass` as the first program pass in this PR.

- Link to the first PR: [[CINN Zero-Dim] CINN supports 0D-Tensor with trick temporarily](https://github.com/PaddlePaddle/Paddle/pull/53382)

### CODE REVIEW

File: cinn/frontend/decomposer/broadcast.cc [hunk 1]
```diff
@@ -118,10 +118,13 @@ void elementwise_add(const Instruction& instr, const DecomposerContext& context)
...
   CHECK_EQ(instr->outputs.size(), 2UL) << "2 output tensors for " << instr->op_type;
-  auto dout     = instr->inputs[0];
-  auto dx       = instr->outputs[0];
-  auto dy       = instr->outputs[1];
-  int axis      = instr.GetAttrs<int>("axis");
+  auto dout = instr->inputs[0];
+  auto dx   = instr->outputs[0];
+  auto dy   = instr->outputs[1];
+  int axis  = instr.GetAttrs<int>("axis");
+  if (axis < 0 && dx->shape.size() < dy->shape.size()) {
```
Comment (thisjiang) [inline cinn/frontend/decomposer/broadcast.cc:125]: 这儿改为`dx->shape.size() < dy->shape.size()`时`dx->shape.size() - dy->shape.size()`可以么？

**Review by thisjiang — COMMENTED** [review]
在python/tests/passes目录下加个python端的Pass测试吧

File: cinn/frontend/decomposer/broadcast.cc (hunk 1 above)
Comment (jiahy0825) [reply cinn/frontend/decomposer/broadcast.cc:125]: 从 `elementwise_add_grad` 函数后续的逻辑来看，个人理解现在 elementwise_add 尚不支持交换律，得严格要求 `dx.shape.size() > dy.shape.size()`，因此在这里添加了报错信息

**Comment by jiahy0825** [comment]
> 在python/tests/passes目录下加个python端的Pass测试吧

Done

**Review by thisjiang — APPROVED** [review]
LGTM

**Pull request merged**